src/3D/Camera.cpp

```diff
@@ -110,101 +178,287 @@ bool Camera::ProjectWorldToScreen(const glm::vec3 worldPosition, const glm::vec4
+			_position = (rotationMatrixX * glm::vec4(_position - handPos, 0.0f)) + glm::vec4(handPos, 0.0f);
+
+			glm::mat4x4 rotationMatrixY(1.0f);
+			rotationMatrixY = glm::rotate(rotationMatrixY, pitch, GetRight());
+			_position = (rotationMatrixY * glm::vec4(_position - handPos, 0.0f)) + glm::vec4(handPos, 0.0f);
+
+			_rotation.x -= pitch;
+			_rotation.y -= yaw;
+		}
 	}
-	else if (e.key.keysym.scancode == SDL_SCANCODE_A)
+	else if (e.type == SDL_MOUSEMOTION && e.motion.state & SDL_BUTTON(SDL_BUTTON_LEFT))
 	{
-		_dv.x += (e.type == SDL_KEYDOWN) ? -1.0f : 1.0f;
+		_mouseIsMoving = true;
 	}
-	else if (e.key.keysym.scancode == SDL_SCANCODE_D)
+	else if (e.type == SDL_MOUSEBUTTONUP && e.button.button == SDL_BUTTON(SDL_BUTTON_LEFT))
 	{
-		_dv.x += (e.type == SDL_KEYDOWN) ? 1.0f : -1.0f;
+		_dv = glm::vec3(0.0f);
+		_mouseIsDown = false;
 	}
-	else if (e.key.keysym.scancode == SDL_SCANCODE_LCTRL)
+	else if (e.type == SDL_MOUSEBUTTONDOWN && e.button.button == SDL_BUTTON(SDL_BUTTON_LEFT))
 	{
-		_dv.y += (e.type == SDL_KEYDOWN) ? -1.0f : 1.0f;
+		_mouseIsDown = true;
 	}
-	else if (e.key.keysym.scancode == SDL_SCANCODE_SPACE)
+	if (e.type == SDL_MOUSEWHEEL)
 	{
-		_dv.y += (e.type == SDL_KEYDOWN) ? 1.0f : -1.0f;
+		auto movementSpeed = _movementSpeed * std::max(log(_position.y), 0.0f) + 1.0f;
+		if (e.wheel.y != 0) // scroll up or down
+		{
+			_velocity.z +=
+			    (((movementSpeed * e.wheel.y * abs(e.wheel.y) * 5 * _maxMovementSpeed) - _velocity.z) * _accelFactor);
+		}
+		if (e.wheel.x != 0) // mouse wheel left or right
+		{
+			_rotation.z += (((movementSpeed * e.wheel.x * _maxMovementSpeed) - _velocity.z) * _accelFactor);
+		}
 	}
 }
 
-void Camera::handleMouseInput(const SDL_Event& e)
+void Camera::Update(std::chrono::microseconds dt)
 {
-	// Holding down the middle mouse button enables free look.
-	if (e.type == SDL_MOUSEMOTION && e.motion.state & SDL_BUTTON(SDL_BUTTON_MIDDLE))
-	{
-		glm::vec3 rot = GetRotation();
-
-		rot.y -= e.motion.xrel * _freeLookSensitivity * 0.1f;
-		rot.x -= e.motion.yrel * _freeLookSensitivity * 0.1f;
+	auto airResistance = .92f; // reduced to make more floaty
+	float fdt = float(dt.count());
+	glm::mat3 rotation = glm::transpose(GetViewMatrix());
 
-		SetRotation(rot);
+	// deal with hand pulling camera around
+	glm::ivec2 handScreenVec(0);
+	float handDist = 0.0f, worldHandDist = 0.0f;
+	int sWidth, sHeight;
+	Game::instance()->GetWindow()->GetSize(sWidth, sHeight);
+	if (_mouseIsDown)
+	{
+		// get hand transform and project to screen coords
+		auto& entityReg = Game::instance()->GetEntityRegistry();
+		auto handEntity = Game::instance()->GetHand();
+		auto& handTransform = entityReg.Get<ecs::components::Transform>(handEntity);
+		auto handPos = handTransform.position;
+		glm::vec3 handToScreen;
+		glm::vec4 viewport = glm::vec4(0, 0, sWidth, sHeight);
+		glm::vec3 hit;
+		if (ProjectWorldToScreen(handPos, viewport, handToScreen) and RaycastMouseToLand(hit))
```

Comment:
This isn't Python ;)

Use `&&`

Reply:
there's a python wrapper for BGFX maybe we should switch the entire thing to python so I can use `and`